tools/clang/test/CodeGenDXIL/hlsl/intrinsics/buffer-agg-load-stores.hlsl

```diff
@@ -0,0 +1,231 @@
+
+
+
+// CHECK: %dx.types.ResRet.[[TY:[a-z][0-9][0-9]]] = type { [[TYPE:[a-z0-9]*]],
+
+#if defined(ARR)
+#define TYNAME(T,N) T[N]
+#define DECL(T,N,V) T V[N]
+#elif defined(MAT)
+#define TYNAME(T,N) matrix< T, N, N >
+#define DECL(T,N,V) matrix< T, N, N > V
+#elif defined(OFF)
+#define TYNAME(T,N) OffVector< T, N >
+#define DECL(T,N,V) OffVector< T, N > V
+#else
+#define TYNAME(T,N) Vector< T, N >
+#define DECL(T,N,V) Vector< T, N > V
+#endif
+
+template<typename T, int N>
+struct Vector {
+  vector<T, N> v;
+  Vector operator+(Vector vec) {
+    Vector ret;
+    ret.v = v + vec.v;
+    return ret;
+  }
+};
+
+template<typename T, int N>
+struct OffVector {
+  float4 pad1;
+  double pad2;
+  vector<T, N> v;
+  OffVector operator+(OffVector vec) {
+    OffVector ret;
+    ret.pad1 = 0.0;
+    ret.pad2 = 0.0;
+    ret.v = v + vec.v;
+    return ret;
+  }
+};
+
+  ByteAddressBuffer RoByBuf : register(t1);
+RWByteAddressBuffer RwByBuf : register(u1);
+
+  StructuredBuffer< TYNAME(TYPE,NUM) > RoStBuf : register(t2);
+RWStructuredBuffer< TYNAME(TYPE,NUM) > RwStBuf : register(u2);
+
+ConsumeStructuredBuffer< TYNAME(TYPE,NUM) > CnStBuf : register(u4);
+AppendStructuredBuffer< TYNAME(TYPE,NUM) > ApStBuf  : register(u5);
+
+TYPE Add(TYPE f1[NUM], TYPE f2[NUM])[NUM] {
+  TYPE ret[NUM];
+  for (int i = 0; i < NUM; i++)
+    ret[i] = f1[i] + f2[i];
+  return ret;
+}
+
+template<typename T>
+T Add(T v1, T v2) { return v1 + v2; }
+
+TYPE Add(TYPE f1[NUM], TYPE f2[NUM], TYPE f3[NUM], TYPE f4[NUM])[NUM] {
+  TYPE ret[NUM];
+  for (int i = 0; i < NUM; i++)
+    ret[i] = f1[i] + f2[i] + f3[i] + f4[i];
+  return ret;
+}
+
+template<typename T>
+T Add(T v1, T v2, T v3, T v4) { return v1 + v2 + v3 + v4; }
+
+void main(uint SIx[2] : SIX, uint1 VIx[2] : VIX) {
+  // ByteAddressBuffer Tests
+
+  // CHECK-DAG: [[HDLROBY:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 1, i32 1, i32 0, i8 0 }, i32 1, i1 false)
+  // CHECK-DAG: [[HDLRWBY:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 1, i32 1, i32 0, i8 1 }, i32 1, i1 false)
+
+  // CHECK-DAG: [[HDLROST:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 2, i32 2, i32 0, i8 0 }, i32 2, i1 false)
+  // CHECK-DAG: [[HDLRWST:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 2, i32 2, i32 0, i8 1 }, i32 2, i1 false)
+
+  // CHECK-DAG: [[HDLCON:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 4, i32 4, i32 0, i8 1 }, i32 4, i1 false)
+  // CHECK-DAG: [[HDLAPP:%.*]] = call %dx.types.Handle @dx.op.createHandleFromBinding(i32 217, %dx.types.ResBind { i32 5, i32 5, i32 0, i8 1 }, i32 5, i1 false)
+
+  // CHECK-DAG: [[IX0:%.*]] = call i32 @dx.op.loadInput.i32(i32 4, i32 {{[0-9]*}}, i32 [[BOFF:0]]
+  // CHECK-DAG: [[RIX0:%.*]] = call i32 @dx.op.loadInput.i32(i32 4, i32 {{[0-9]*}}, i32 [[BOFF]]
```

Comment:
Another pattern of CHECK-DAG where BOFF is defined by the first one and used by the second, and they would both match exactly the same IR (unless I'm missing something), are these mixed in with the create handle ops above them?  If they could match the same thing, how do you know if IX0 or RIX0 are what you want elsewhere?

This test seems pretty complicated and might be difficult to update or maintain if that was necessary.

Reply:
That's correct. The intent is to initially populate RIX0 with the same thing as IX0 for the cases where there is no offset, so it won't get set to anything else. None of this would be necessary with modern FileCheck that allows defining variables through an invocation flag. Barring that, this is what I can do.

I'm not sure how to respond to the concern about the test being complicated. I could add some comments to try to make the function clearer. I suppose much of the complication comes from the weird way arrays are used. I could pare that part off, but all the checks would be the same and any updating would need to be in both places, which is also a maintenance burden that often gets overlooked.

Reply:
I simplified the defines and macros used in this test. Hopefully that will make this more maintainable?